### ISSUE TITLE
Rewrite the VM bytecode interpreter in C?

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#58]
Should we rewrite the VM bytecode interpreter from assembler to C? Let me make the case for consideration:

The assembler VM has a high maintenance cost and a low benefit.

High maintenance cost because the code is complex and the maintenance complexity scales poorly, perhaps O(N log N) when supporting N CPUs (i.e. maintaining a separate assembler VM for each architecture + keeping them perfectly synchronized.) The main reason that all non-essential RaptorJIT platforms were removed is not having maintainer resources to take care of all their VMs.

The benefit is low because RaptorJIT targets server applications that probably spend ~0.1% of their time running interpreted code after initialization and warm-up. The other 99.9% of the time is spent in JIT mcode, the runtime system, and FFI functions none of which benefit from the assembler VM.

So it seems reasonable to me to consider porting the bytecode interpreter to C if this will make the maintenance much simpler, and would change the economics of supporting multiple CPU architectures (maintain JIT backend only), and would only cost perhaps <0.1% in total real-world performance.

(Just now this issue is not urgent because we only support one architecture and maintaining one VM is okay. However, if we want to add support for more architectures like ARM64 then we will have to decide whether we will maintain the VMs as parallel assembler programs or write a common VM in C.)

**Comment by wingo** [comment]
One benefit of having an interpreter in assembly is that you control the stack and you can tail-call between the interpreter and compiled traces.  If you are looking at simplifying, I would consider instead compiling the interpreter from the IR used by the optimizing compiler.  I.e. each instruction has its associated code generator, coded in C, and you run (a component of) RaptorJIT during the build to generate the assembly for the architecture you are targeting.

**Comment by wingo** [comment]
Also having the interpreter in assembly has GC benefits (known stack layout).  Anyway to me the tradeoff does not go obviously towards the C direction (though it might be doable as you say).

**Comment by lukego** [comment]
@wingo Could be that porting to C is the wrong idea. Questions though: Can you force tail-calls in C with clang 4.0+? Would the stack layout issues disrupt the operation of the current GC?

**Comment by wingo** [comment]
I do not know about "forcing".  It could be that it works but I don't think you have any guarantee.  I could be wrong though.

Of course you can use other trace-transition strategies like trampolining or something, that's a possibility too.

**Comment by lukego** [comment]
Complex problem :).

Sounds like it makes sense to continue enjoying the lovely current situation of having the VM in assembler and only supporting x64 unless/until somebody comes along with a better solution.

I am certainly not in a hurry to look under the hood of the C compiler. I looked at the LLVM sources for the first time in the context of #55 and that made me very grateful for the work Mike has done on the simple code generator in LuaJIT. It's a bit cryptic but boy am I not in a hurry to dig into millions of lines of C++ :).

**Comment by wingo** [comment]
An idle thought: perhaps generate the interpreter from Lua using DynASM?  You check in the built product to git to avoid bootstrapping problems.  Could be fun!  Could allow you to parameterize the interpreter in different ways.

**Comment by lukego** [comment]
@wingo Yes I would like to switch the VM over to the Lua DynASM (#13).

I feel like a pragmatic approach to the bootstrapping problem is to allow well-defined dependencies at build time and make nix responsible for providing them. Currently I am including LuaJIT 2.0 as a build dependency and this allowed me to remove the `minilua` stripped-down Lua interpreter from LuaJIT (#28). LuaJIT needs a working bootstrap Lua at build time because even though they use the C-based DynASM they still need Lua to run the DynASM preprocessor...

Could also use PUC Lua for bootstrap instead of LuaJIT. I picked LuaJIT 2.0 because it already includes the `bit` module and doesn't need the add-on module installed. (It was already packaged with nixpkgs too.)

**Comment by lukego** [comment]
Generally I feel like nix completely changes the build dependencies problem. We are able to depend on a lot of different tools and nix will avoid "dependency hell" by pinning them all to exact versions (sha256 hash of their sources) and automatically installing them. This means we should also be able to pull in the Rust-based DynASM instruction encoding generator from #31 when that is ready.

RaptorJIT is still being equally conservative about runtime dependencies as LuaJIT though...

**Comment by lukego** [comment]
(I realize there is a certain charm to keeping dependencies small and not needing a tool like nix, but I also feel that it more than pays for itself quickly. I am especially keen for everybody to be able to run all the standard benchmark visualizations #59 and I _know_ that nobody is going to manually install all the right versions of all the R library dependencies to do that :-))

**Comment by wingo** [comment]
Using LuaJIT as a build-dep makes sense to me too, yeah.  In the Snabb case anyway.  I only wonder about source distribution -- right now Snabb has a nice story, just check out and make; will source distributions now bundle both LuaJIT and RaptorJIT?  It's possible of course.  For me being able to make a basic build on whatever system is quite valuable.

**Comment by javierguerragiraldez** [comment]
the interpreter is actually written in DynASM's assembler.  check any of the `vm_xxxx.dasc` files, there's a `build_subroutines()` and a `build_ins()` functions.  the first one creates most of the "utility" functions, while the second is called repeatedly to fill the `dispatch` table with each instruction's implementation.

**Comment by lukego** [comment]
@wingo Good topic. I see two extreme positions and one middle ground.

One extreme is to keep build dependencies to a bare minimum e.g. a random version of make+gcc+shellutils+binutils. This is convenient, familiar, and not terribly robust. LuaJIT does this and Snabb preserves it. Downsides: you have to jump through hoops to avoid dependencies (#28), surprise requirements can slip in (e.g. uses of find/grep/awk/etc in the Makefile snabbco/snabb#510), and surprising aspects of the build environment can turn up in the binary e.g. C compilers with different optimizations and warnings, dynamic linking to specific GLIBC symbol versions from the build host (snabbco/snabb#359), etc.

The other extreme is to build with `nix-build`. This is semi-convenient, weird and unfamiliar, and very robust. Upside is that you can use all the tools you want and it will build _exactly_ the same way on any Linux host. Downside is that you need to [read about nix](https://www.infoq.com/articles/configuration-management-with-nix) before it makes sense and you have to wait for it to download all the tools. (Similar situation to having Docker as a build dependency.)

One middle ground could be to have three kinds of dependencies: runtime dependencies (few as possible), build dependencies (minimal like LuaJIT), and bootstrap dependencies (build with nix and checked into the source tree.) This way as an end-user you can simply do `make` while VM hackers will use nix for e.g. rebuilding the VM using the bootstrap version of LuaJIT to run DynASM, or regenerating the DynASM instruction list using Intel XED definitions, etc.

**Comment by wingo** [comment]
Testing out this reply-by-email functionality :)

On Tue 28 Mar 2017 15:45, Luke Gorrie <notifications@github.com> writes:

> @wingo Good topic. I see two extreme positions and one middle ground.

Sounds about right to me.  It's hard to find the right "simple" -- I
mean all three of the options you mention are simple in their own way.
And I could add a fourth: just vendoring in more things, like bundling
LuaJIT and RaptorJIT source with Snabb.  Prolly not a great idea but
it's possible and simple in its own way (from user POV).

I am going to make a first-person perspective now :)  By no means let
this constrain what you do.  Just some input:

I think it's important to be able to easily rebuild any bootstrap
binaries that we include in Snabb.

I would like to avoid *having* to use nix-build to build Snabb, or "guix
build" for that matter.  I feel like I need the console build
environment.

I think it would be nice if the set of bootstrap dependencies that Snabb
has is fairly common.  Like it should be possible to use them from
Ubuntu if that's your thing.  We can require an apt-get install whatever
but if we start to use less common deps for common things then I feel
like we are getting into trouble.  I'm very happy to use R via Nix /
Guix though; R is complicated :P

I think these things point towards Snabb's build just searching for
needed bootstrap binaries in $PATH, having added
$top_srcdir/bootstrap-deps to $PATH.  If someone deletes
$top_srcdir/bootstrap-deps then Snabb will find e.g. "luajit" in $PATH
and it's the user's problem.

FWIW my ideal dev environment is "guix environment --load ./package.scm
snabb".  This gives me a shell in which all dependencies of the "snabb"
package are available (like nix-env).  In this case ./package.scm is
ideally checked into Snabb git and defines a Guix package for Snabb and
any possibly-unpackaged dependencies.  We should have a similar file for
Nix too, I would imagine; dunno tho, people can be on different channels
in Nix; Guix currently just has one channel.

How does this fit with where you are going?

Andy

**Comment by lukego** [comment]
@wingo So also first-person perspective :)

Sweet spot seems to me to be:
- Snabb "vendors in" the source for RaptorJIT (with `git subtree` as we do for LuaJIT now.)
- Snabb and RaptorJIT can each build with `make` with standard tools (no nix.)
- RaptorJIT uses nix for bootstrapping & commits the generated sources.
- Application developers can further lock down the build environment using nix, guix, docker, etc. This is self-supported downstream work.

Note that at the moment I have [pinned on one specific version of `nixpkgs`](https://github.com/raptorjit/raptorjit/blob/master/pkgs.nix) and this implies the exact version of all tools in the toolchain. If you use a Guix definition instead then you will get another set of versions instead. I would say this is fine: on the one hand RaptorJIT has a well-defined toolchain for bootstrapping, on the other hand application developers can use a similar-but-different one if that is what they prefer. (Just like right now you and I are both shipping `snabb` binaries and we are each building them using our own preferred toolchains and supporting them accordingly.)

**Comment by wingo** [comment]
All 4 points sound great to me.  I am on board :)

**Comment by lukego** [comment]
@wingo Groovy!

On reflection one more category is _testing_ dependencies. I reckon this is an area where we want to leverage a lot of useful tools and it is helpful to capture those dependencies with nix. That way everybody can run all of the tests - including the CI servers.

If our test suites end up depending on R, LOOM, Valgrind, afl-fuzz, QuickCheck, QEMU, etc, then that seems like a good problem to have. Probably there will be a division of labor with some people knowing specific testing tools and other people helping them to wrap them with nix (which tends to be fairly easy but takes some practice.)

**Comment by iehrlich** [comment]
Quickly glanced through all replies, found only one mention of bootstrapping - we need another one here! :)

You can strip the Lua/LuaJIT dependency for DynASM, if you
1) prepare pure-C main loop;
2) compile everything and this main loop with C compiler, getting working (yet slower) build of RaptorJIT
3) use this build to generate asm main loop using DynASM;
4) re-link already compiled sources with asm main loop, instead of C main loop
5) PROFIT!

Basically, you'll get two pluggable equally working main loops - pure-C portable but slower one, and its assembler counterpart. Also, 1st gear will be fully portable, if JIT is disabled. 

Also, having these two builds, you'll be able to (a) validate both and (b) compare performance continuously.

**Comment by lukego** [comment]
@iehrlich I suppose it is very important to be clear about goals. One person's "profit" is somebody else's "loss" :-).

RaptorJIT needs above all to be easy to maintain. This is to attract maintainers and focus their time on the most productive aspects of the project. This means making the build process as simple as possible without sacrificing practicality.

I hope the sweet spot will be for end-users to only need basic stone tools (e.g. make and C compiler) while VM hackers can assume the existence of more tools such as a Lua interpreter for regenerating the VM sources.

I understand the urge to geek out on bootstrapping. It's a fascinating topic. I can also imagine that for some projects having a sexy bootstrapping story helps to attract good developers. However, for RaptorJIT it is important to keep things simple, and I feel like there are already ample vehicles for geeking out on bootstrapping in the world e.g. [jonesforth](https://github.com/nornagon/jonesforth/blob/master/jonesforth.S), [nand2tetris](http://www.nand2tetris.org/), [Reflections on Trusting Trust](https://www.ece.cmu.edu/~ganger/712.fall02/papers/p761-thompson.pdf), SICP, etc. Those projects let you go very deep down the rabbit hole without creating maintenance work for other people :).

I am a bit sad that the only [talk](https://www.youtube.com/watch?v=8Q0KLTma_FA) about LuaJIT by Peter Cawley needed 42 minutes just to explain the build system. There are _so_ many other things that I would love to heard him talk about in that time instead. But that's me.

**Comment by lukego** [comment]
I also think that Nix is a big idea that deserves to influence the bootstrapping landscape since 2006 when Eelco Dolstra published his PhD thesis [The Purely Functional Software Deployment Model](https://nixos.org/~eelco/pubs/phd-thesis.pdf). Now _that_ is a practical topic to geek out on in this day and age :-) dependencies without dependency hell.

**Comment by lukego** [comment]
> Also, having these two builds, you'll be able to (a) validate both and (b) compare performance continuously.

This is an interesting point that I missed the first time around. Yes, it's a very interesting testing technique to have one simple slow implementation and one fast complex one, and then to run automatic tests looking for differences is behavior that need to be checked as potential bugs.

Is this done much in practice? I often compare two different verisons of the same program in this way but that's not quite the same thing. This does remind me a little of QuickCheck where you define a detailed model of your program and look for contradictions from running the actual software. Also perhaps related to [Squeak](http://files.squeak.org/docs/OOPSLA.Squeak.html) where they write the VM in a subset of their own language ("slang") which is simple enough to be automatically translated to C. So they can develop the VM using high level interactive tools, and generate fast C code for the VM, and check the translator by comparing the interpreted vs compiled VMs. Cool stuff.

But now I am going down that rabbit hole...... :-)

**Comment by iehrlich** [comment]
> RaptorJIT needs above all to be easy to maintain. This is to attract maintainers and focus their time on the most productive aspects of the project. This means making the build process as simple as possible without sacrificing practicality.
On the scale from a bash script to LLVM, LuaJIT (and, consequently, RaptorJIT) is quite a complex piece of software. It will definitely be hard to understand and maintain, but the goal you might consider is removing _unnecessary_ complexity, as there's plenty of it. In the same time, the bootstrapping idea is quite common, it yields known benefits, and it allows to strip a dependency.

Another nice idea is to transition from DynASM itself, because it's not quite necessary for VM, but that's not the point :)

> This is an interesting point that I missed the first time around. Yes, it's a very interesting testing technique to have one simple slow implementation and one fast complex one, and then to run automatic tests looking for differences is behavior that need to be checked as potential bugs.
> 
> Is this done much in practice?
> 
Short answer is yes. Unfortunately, I don't have any real-life open-source example right now, but this is definitely done in the industry, when it comes to this sort of "geared" systems.

As a side-note, it's quite an interesting task to build a homogeneous stability and performance validation engine, that will run the same tests on (a) pure-C main loop w/o JIT, (b) asm main loop w/o JIT, (c) asm main loop w/ JIT. Something like that.

**Comment by dibyendumajumdar** [comment]
Hi, I discovered this project today! 

I was interested in porting the LuaJIT VM to C - and asked a question regarding this some time ago!

https://www.freelists.org/post/luajit/Luajit-interpreter-question

I think now that it will probably run much slower because part of the reason for LuaJIT interpreter's performance is the tuned branching which is probably impossible to replicate in C.

**Comment by lukego** [comment]
Thanks for that relevant link to past discussion, @dibyendumajumdar!

**Comment by dibyendumajumdar** [comment]
LuaJIT's interpreter is almost twice as fast as the Lua interpreter - I played around with some tweaks to the Lua interpreter to see if its performance could be improved to the level of LuaJIT. Sadly I wasn't able to. It seems that the performance achieved by LuaJIT is very much the result of hand optimised register allocation and branching - neither can be replicated in C.

**Comment by lukego** [comment]
@dibyendumajumdar I am recommending against ever allowing code to run interpreted in the RaptorJIT Optimization Guide (#68). The interpreter may be fast _for an interpreter_ but at least for Snabb we have found it much too slow compared with compile code. This is both due to the extra work of interpretation and also the lack of key optimizations like allocation sinking leading to excessive garbage collection.

My holistic view is that we should minimize the engineering time that we sepnd on the interpreter. That is because we can impact practical application performance much more in other ways: improving JIT, improving profiling, improving performance tuning documentation, etc. So optimizing the interpreter implementation for simplicity over performance makes sense to me (at least when we are talking about a factor of 2 difference.)

Having said that: Maintaining the current x86-64/GC64 VM as an assembler program does seem to be the most practical path for RaptorJIT and the path of least resistance. And quite pleasant too because it's a very nicely written program and a pleasure to work with.
